#### Gini Impurity Implementation

In [1]:
import numpy as np
import pandas as pd

##### Define gini impurity and weighted gini functions

In [3]:
def gini_impurity(class_counts):
    """
    Calculate Gini inpurity for multi_class problem

    Arg:
        calss_counts: array of counts for each class
    
    Return:
        Gini impurity value
    """

    total = np.sum(class_counts)
    proporation = class_counts / total

    # Method 1: 1 - np.sum(p_i^2)
    gini = 1 - np.sum(proporation ** 2)

    return gini

def weighted_gini(left_counts, right_counts):
    """
    Calculate weighted Gini impurity after a split
    """

    left_total = np.sum(left_counts)
    right_total = np.sum(right_counts)
    total = left_total + right_total

    gini_left = gini_impurity(left_counts)
    gini_right = gini_impurity(right_counts)

    weight = (left_total / total) * gini_left + (right_total / total) * gini_right

    return weight

In [8]:
# parent gini
parent_counts = np.array([80, 60, 40, 20])
gini_parent = gini_impurity(parent_counts)
print(f'Parent Gini: {gini_parent:.4f}')

Parent Gini: 0.7000


In [9]:
# after split
left_counts = np.array([75, 15, 8, 2])
right_counts = np.array([5, 45, 32, 18])

gini_left = gini_impurity(left_counts)
gini_right = gini_impurity(right_counts)
gini_weighted = weighted_gini(left_counts, right_counts)

print(f'Left Gini: {gini_left:.4f}')
print(f'Right Gini: {gini_right:.4f}')
print(f'Weighted_gini: {gini_weighted:.4f}')

gini_gain = gini_parent - gini_weighted
print(f'Gini Gain: {gini_gain:.4f}')

Left Gini: 0.4082
Right Gini: 0.6602
Weighted_gini: 0.5342
Gini Gain: 0.1658


# Gini Impurity for Multi-Class Problems

Let me show you step-by-step how to calculate Gini impurity when you have more than 2 classes.

---

## The Formula

$$\text{Gini}(S) = 1 - \sum_{i=1}^{c} p_i^2$$

Where:
- **S** = dataset/node
- **c** = number of classes
- **p_i** = proportion of class i in the dataset

Or equivalently (the misclassification interpretation):

$$\text{Gini}(S) = \sum_{i=1}^{c} p_i(1 - p_i)$$

---

## Example 1: 3-Class Problem

**Dataset: 100 emails**
- 50 Work
- 30 Personal
- 20 Spam

### Calculate Proportions:
- p₁ (Work) = 50/100 = 0.50
- p₂ (Personal) = 30/100 = 0.30
- p₃ (Spam) = 20/100 = 0.20

### Method 1: Using 1 - Σp²

$$\text{Gini} = 1 - (p_1^2 + p_2^2 + p_3^2)$$

$$\text{Gini} = 1 - (0.50^2 + 0.30^2 + 0.20^2)$$

$$\text{Gini} = 1 - (0.25 + 0.09 + 0.04)$$

$$\text{Gini} = 1 - 0.38 = 0.62$$

### Method 2: Using Σp(1-p) [Misclassification]

$$\text{Gini} = p_1(1-p_1) + p_2(1-p_2) + p_3(1-p_3)$$

$$\text{Gini} = 0.50(0.50) + 0.30(0.70) + 0.20(0.80)$$

$$\text{Gini} = 0.25 + 0.21 + 0.16$$

$$\text{Gini} = 0.62$$

✓ Both methods give the same answer!

---

## Example 2: 4-Class Problem (Email Categories)

**Dataset: 200 emails**
- 80 Work
- 60 Personal
- 40 Promotions
- 20 Spam

### Calculate Proportions:
- p₁ (Work) = 80/200 = 0.40
- p₂ (Personal) = 60/200 = 0.30
- p₃ (Promotions) = 40/200 = 0.20
- p₄ (Spam) = 20/200 = 0.10

### Method 1: Using 1 - Σp²

$$\text{Gini} = 1 - (0.40^2 + 0.30^2 + 0.20^2 + 0.10^2)$$

$$\text{Gini} = 1 - (0.16 + 0.09 + 0.04 + 0.01)$$

$$\text{Gini} = 1 - 0.30 = 0.70$$

### Method 2: Using Σp(1-p)

$$\text{Gini} = 0.40(0.60) + 0.30(0.70) + 0.20(0.80) + 0.10(0.90)$$

$$\text{Gini} = 0.24 + 0.21 + 0.16 + 0.09$$

$$\text{Gini} = 0.70$$

✓ Same answer!

---

## Interpreting the Result

**Gini = 0.70 means:**

If you randomly pick an email and randomly assign it a class label (using the node's distribution), you have a **70% chance of guessing wrong**.

### Breaking it down by class:
- **Work (40%)**: 40% chance you pick it, 60% chance you guess wrong → 0.24 misclassification
- **Personal (30%)**: 30% chance you pick it, 70% chance you guess wrong → 0.21 misclassification
- **Promotions (20%)**: 20% chance you pick it, 80% chance you guess wrong → 0.16 misclassification
- **Spam (10%)**: 10% chance you pick it, 90% chance you guess wrong → 0.09 misclassification

**Total risk = 0.70**

---

## Example 3: Pure Node vs. Impure Node

### Pure Node (All one class)
**Dataset: 100 emails, ALL Work**
- p₁ (Work) = 1.0
- p₂ (Personal) = 0.0
- p₃ (Promotions) = 0.0
- p₄ (Spam) = 0.0

$$\text{Gini} = 1 - (1.0^2 + 0.0^2 + 0.0^2 + 0.0^2)$$

$$\text{Gini} = 1 - 1.0 = 0$$

**Interpretation:** 0% chance of misclassification. Perfect purity!

### Balanced Node (Equal distribution)
**Dataset: 100 emails, 25 of each class**
- p₁ = p₂ = p₃ = p₄ = 0.25

$$\text{Gini} = 1 - (0.25^2 + 0.25^2 + 0.25^2 + 0.25^2)$$

$$\text{Gini} = 1 - (4 \times 0.0625)$$

$$\text{Gini} = 1 - 0.25 = 0.75$$

**Interpretation:** 75% chance of misclassification. Maximum impurity!

---

## Gini for Child Nodes (Weighted Gini)

When you split a node, you need to calculate the **weighted average Gini** of the child nodes.

### Formula:

$$\text{Gini}_{\text{weighted}} = \frac{|S_{\text{left}}|}{|S|} \times \text{Gini}(S_{\text{left}}) + \frac{|S_{\text{right}}|}{|S|} \times \text{Gini}(S_{\text{right}})$$

### Example: Split on "Sender Domain"

**Parent node: 200 emails**
- 80 Work, 60 Personal, 40 Promo, 20 Spam
- Gini(parent) = 0.70

**Split: Sender Domain = work.com?**

**Left child (work.com): 100 emails**
- 75 Work, 15 Personal, 8 Promo, 2 Spam
- p₁ = 0.75, p₂ = 0.15, p₃ = 0.08, p₄ = 0.02

$$\text{Gini}_{\text{left}} = 1 - (0.75^2 + 0.15^2 + 0.08^2 + 0.02^2)$$

$$= 1 - (0.5625 + 0.0225 + 0.0064 + 0.0004)$$

$$= 1 - 0.5918 = 0.4082$$

**Right child (other domains): 100 emails**
- 5 Work, 45 Personal, 32 Promo, 18 Spam
- p₁ = 0.05, p₂ = 0.45, p₃ = 0.32, p₄ = 0.18

$$\text{Gini}_{\text{right}} = 1 - (0.05^2 + 0.45^2 + 0.32^2 + 0.18^2)$$

$$= 1 - (0.0025 + 0.2025 + 0.1024 + 0.0324)$$

$$= 1 - 0.3398 = 0.6602$$

**Weighted Gini:**

$$\text{Gini}_{\text{children}} = \frac{100}{200} \times 0.4082 + \frac{100}{200} \times 0.6602$$

$$= 0.5 \times 0.4082 + 0.5 \times 0.6602$$

$$= 0.2041 + 0.3301 = 0.5342$$

### Gini Gain:

$$\text{Gain} = \text{Gini}_{\text{parent}} - \text{Gini}_{\text{children}}$$

$$\text{Gain} = 0.70 - 0.5342 = 0.1658$$

This split **reduces misclassification risk by 16.58%**. Good split! ✓

---

## Key Formulas Summary

| Scenario | Formula | Example |
|----------|---------|---------|
| **Single node Gini** | $1 - \sum p_i^2$ | $1 - (0.4^2 + 0.3^2 + 0.2^2 + 0.1^2) = 0.70$ |
| **Misclassification form** | $\sum p_i(1-p_i)$ | $0.4(0.6) + 0.3(0.7) + 0.2(0.8) + 0.1(0.9) = 0.70$ |
| **Weighted (after split)** | $\frac{\|L\|}{\|S\|}Gini(L) + \frac{\|R\|}{\|S\|}Gini(R)$ | $0.5(0.41) + 0.5(0.66) = 0.53$ |
| **Gini Gain** | $Gini(parent) - Gini(children)$ | $0.70 - 0.53 = 0.17$ |

---

Does this clear up how to calculate Gini impurity for multiple classes? Want me to show how scikit-learn uses this internally, or dive into comparing different splits?